source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -109,11 +109,27 @@ LookupContextPtr SimpleHttpCache::makeLookupContext(LookupRequest&& request) {
   return std::make_unique<SimpleLookupContext>(*this, std::move(request));
 }
 
-void SimpleHttpCache::updateHeaders(const LookupContext&, const Http::ResponseHeaderMap&,
-                                    const ResponseMetadata&) {
-  // TODO(toddmgreer): Support updating headers.
-  // Not implemented yet, however this is called during tests
-  // NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
+                                    const Http::ResponseHeaderMap& response_headers,
+                                    const ResponseMetadata& metadata) {
+    absl::WriterMutexLock lock(&mutex_);
+    const auto& simple_lookup_context = static_cast<const SimpleLookupContext&>(lookup_context);
+    const Key& key = simple_lookup_context.request().key();
+
+    auto iter = map_.find(key);
+    if (iter == map_.end() || !iter->second.response_headers_) {
+      return;
+    }
+    auto& entry = iter->second;
+
+    // TODO(tangsaidi) handle Vary header updates properly
+    if (VaryHeader::hasVary(*(entry.response_headers_))) {
+      return;
+    }
+
+    entry.response_headers_ =
```

Comment:
We should compare the headers currently in the cache with the ones we received from the backend to make sure they match. 

It's safer to skip the update and send the request to the backend for a full fetch and insert rather than risk mismatched content due to a bug in the upstream service's validation logic.

Reply:
We had a discussion offline. Here's a summary of our discussion:

There are potential problems with updating all the headers,
1. Key conflicts on etags would make them unreliable
2. Incorrect upstream etag usage can lead to errors
3. Race condition with an insert operation

These are real problems we need to consider for a large-scale cache system. For 1, there's a 50% chance of getting a SHA1 collision with 2^80 hashes. This means it's extremely rare even at Google scale. For 2, we would like to make the assumption that the upstream service is correct, at least for our simple HTTP cache. For 3, we would like to make another assumption that we get a response instantly from upstream.

We make these assumptions to avoid over-complicating our simple cache filter example while complying with RFC7234 4.2.4. I've added additional comments in the code to explain the assumptions. 

The implementation in this PR updates all the headers in accordance with RFC7234, with some caveats identified as TODO.